### PR TITLE
Add granite model support

### PR DESCRIPTION
**Pull request by amdrajeevp1**
## Summary

  Adds support for exporting `ibm-granite/granite-4.0-1b` (and compatible granite-4.0 variants) via the OGA model
  builder.

  `granite-4.0-1b` uses the `GraniteMoeHybridForCausalLM` architecture, which was previously unsupported — falling back
  to `GraniteForCausalLM` caused incorrect weight mapping and significant accuracy loss. This PR adds a dedicated
  `GraniteMoeHybridModel` builder that handles the structural differences while reusing existing granite infrastructure.

  **Key differences from GraniteForCausalLM:**
  - MLP weights are in `shared_mlp.input_linear` (fused gate+up) and `shared_mlp.output_linear` (down) instead of
  `mlp.gate_proj/up_proj/down_proj`
  - MLP width comes from `shared_intermediate_size` rather than `intermediate_size`
  - `rope_theta` is stored under a `rope_parameters` dict in the config

  For the 1B variant, all layers are attention-type and MoE routing is inactive (`num_local_experts=0`), so it is
  computationally equivalent to GraniteForCausalLM with only the weight layout differing.

  ## Changes

  - `src/python/py/models/builders/granite.py` — adds `GraniteMoeHybridModel` subclassing `GraniteModel`; overrides
  `make_layer` and `make_mlp_unpacked_regular` to handle the fused input_linear weight and correct intermediate size
  - `src/python/py/models/builders/base.py` — adds `rope_parameters` dict fallback in rope_theta resolution
  - `src/python/py/models/builder.py` — maps `GraniteMoeHybridForCausalLM` → `GraniteMoeHybridModel`
  - `src/python/py/models/builders/__init__.py` — exports `GraniteMoeHybridModel`
  - `src/models/model_type.h` — registers `granitemoehybrid` in `IsLLM()`

  ## Test plan

  - [ ] Export `ibm-granite/granite-4.0-1b-base` with `python builder.py -m ibm-granite/granite-4.0-1b-base -o out/ -p
  fp32 -e cpu`
  - [ ] Verify model loads and generates text without errors
  - [ ] Confirm output matches PyTorch reference
  - [ ] Confirm existing `GraniteForCausalLM` models (granite-3.x) are unaffected

### CODE REVIEW

File: src/python/py/models/builders/granite.py [hunk 1]
```diff
@@ -67,3 +67,47 @@ def make_layer(self, layer_id, layer):
...
+        mlp = layer.shared_mlp
+        mlp.gate_up_proj = mlp.input_linear
+        mlp.down_proj = mlp.output_linear
+        layer.mlp = mlp
+        super().make_layer(layer_id, layer)
+
+    def make_mlp_unpacked_regular(self, layer_id, mlp, gate_up_linear, root_input):
+        # Override to split input_linear at _mlp_intermediate_size explicitly,
+        # making the slicing independent of self.intermediate_size.
+        import torch
```
Comment (kunal-vaishnavi) [inline src/python/py/models/builders/granite.py]: Can we move all imports to be global?

Comment (amdrajeevp1) [reply src/python/py/models/builders/granite.py]: Hi @kunal-vaishnavi  - can you please check if the changes are sufficient to your requirements, thanks.

**Comment by kunal-vaishnavi** [comment]
Thanks for your contribution! Can you also make the following additions for in alphabetical order?

- Add the model type to the [model builder README](https://github.com/microsoft/onnxruntime-genai/blob/main/src/python/py/models/README.md#current-support)
- Add the model type to the [repo's main README](https://github.com/microsoft/onnxruntime-genai/blob/main/README.md)

**Review by kunal-vaishnavi — APPROVED** [review]
